### DISCUSSION TITLE
RateLimiter doesn't wait the full limitRefreshPeriod

### DISCUSSION THREAD

**Discussion by DManstrator** [resilience4j/resilience4j#2182] [Q&A]
Hello. Currently I try to implement throttling via. the `RateLimiter`. I'm accessing a 3rd party REST API which has a RateLimit of 1 req/s. The problem is that I need to make at least two requests in probably less than a second for my desired data.

I configured the RateLimier like this:
```java
private static final RateLimiter EXT_PARTY_THROTTLER = RateLimiterRegistry.of(
        RateLimiterConfig.custom()
                .limitRefreshPeriod(Duration.ofSeconds(1))
                .limitForPeriod(1)
                .timeoutDuration(Duration.ofSeconds(15))  // Ensure the timeout is high enough to not throw an exception.
                .build()
).rateLimiter("extParty");
```

When playing around with it I noticed that both requests were started within a second which should cause an exception when performing the API request.

I guess the reasoning behind this is an unlucky time window? So the cycle starts, 400ms later the first request is made, 600ms pass which finishes the cycle and then the second request is made. So two requests will be send within 600ms. At least this is what I except to be the reason.

Would it be possible to configure the RateLimiter so `limitRefreshPeriod` passes between each request?

Kind regards,

Daniel

**Comment by RobWin** [comment]
Yes exactly. The first cycle starts at the time the RateLimiter is created not at the time the first request is made.

**Reply by RobWin** [reply]
See https://github.com/resilience4j/resilience4j/pull/2100 for a PR which tries to fix it

**Reply by DManstrator** [reply]
I worked around this by creating the RateLimiter just in time.

```java
private final RateLimiterRegistry rateLimiterRegistry;

// Creating a new RateLimiter lazily to nearly guarantee correct waiting time.
// See https://github.com/resilience4j/resilience4j/issues/2064#issuecomment-1828676152.
private io.github.resilience4j.ratelimiter.RateLimiter rateLimiter;  // Initialize lazily.

@Autowired
public RateLimitedClient() {
    rateLimiterRegistry = RateLimiterRegistry.of(
            RateLimiterConfig.custom()
                    .limitRefreshPeriod(Duration.ofSeconds(1))
                    .limitForPeriod(1)
                    .timeoutDuration(Duration.ofSeconds(15))  // Ensure the timeout is high enough to not throw an exception.
                    .build()
    );
}

private <R> R createClientAndMakeRequest(final String requestUrl,
                                         final Class<R> responseClass)
        throws Throwable {
    final RestClient restClient = createRestClient(requestUrl);

    final CheckedSupplier<R> throttledCall = io.github.resilience4j.ratelimiter.
            RateLimiter.decorateCheckedSupplier(getRateLimiter(), () ->
                    executeRestClient(restClient, responseClass));

    return throttledCall.get();
}

private io.github.resilience4j.ratelimiter.RateLimiter getRateLimiter() {
    if (rateLimiter == null) {
        rateLimiter = rateLimiterRegistry.rateLimiter("extParty");
    }

    return rateLimiter;
}
```
(Using the full qualified name due to also using the RateLimiter annotation.)

This worked as expected so far.
